EIPS/eip-5593.md

```diff
@@ -0,0 +1,83 @@
+status: Draft
+type: Standards Track
+category: Interface
+created: 2022-09-05
+requires: 1193
+---
+
+## Abstract
+
+Historically the web platform has had a notion of “powerful” APIs like [geolocation](https://www.w3.org/TR/geolocation/) and [camera/microphone](https://www.w3.org/TR/mediacapture-streams/), which are subject to additional security restrictions such as those defined by [secure contexts](https://www.w3.org/TR/secure-contexts/). Since the web3 provider object APIs allow dApp websites to request access to sensitive user data and to request use of user funds, new web3 provider object APIs generally should align to the security considerations of other powerful APIs like geolocation.
+
+## Motivation
+
+Wallets are oftentimes maintaining security and safety of users' funds that can be equivalent to large portions of money. For this reason, it's a good idea to restrict access to the APIs that grant this access in certain contexts to align it more closely with the principles followed for web platform APIs given web3 provider object APIs are often extending this platform via extensions. By adding in restrictions to where extensions inject their APIs, we're reducing the surface area that malicious web pages could fingerprint or falsely request access to the web3 provider object APIs to harm users by accessing their data or stealing their web3 assets.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Restrictions for providers
+
+The provider objects, e.g. `window.ethereum`, are not provided in all contexts. The following restrictions are REQUIRED for conformant wallets:
+
+- Provider objects SHOULD NOT be accessible in private (incognito) windows.
+- The origin MUST be a [secure origin](https://www.chromium.org/Home/chromium-security/prefer-secure-origins-for-powerful-new-features/#definitions) to have access to `window.ethereum`. This can be checked using `window.isSecureContext`, including inside iframes.
+    - Secure contexts include sites that are served from HTTPS but also HTTP `localhost`.
+- By default the web3 provider object APIs MUST NOT be exposed to 3p iframes.
+- `window.ethereum` MUST be `undefined` in an iframe where `window.isSecureContext` returns `false` in that iframe.
+- If the iframe is first-party to the top-level origin AND the `sandbox` attribute is set on the iframe, the provider object MUST be left `undefined`. If `sandbox="allow-same-origin"` attribute is set the provider object MAY be injected. 
+    - Note `"allow-same-origin"` does nothing if the iframe is third-party.
+- If the iframe is a third party to the top-level secure origin, it SHOULD be blocked. It MAY be unblocked if  the iframe uses the `allow=ethereum` attribute. In order to achieve this implementers are REQUIRED to extend the Permissions API.
```

Comment:
I think this violates the spirit of why you block access to the wallet to 3p iframes. If the user is on a malicious website and trying to get the vicitm to do a "click jack" attack. 

For example the user is on a copycat domain and the site to keep the user believing that they want to avoid the connect prompt. They get the user to click on a purchase button or transfer button on the real website through opaque iframes.

Reply:
So test cases for that would be:

https://example.com is granted, a new grant would be required on
https://www.example.com
https://sub.example.com
https://example2.com
http://example.com -- blocked not secure

Reply:
So this would only work if the user grants permission to the ethereum permission via the 3rd party. Most extension based wallets don't get access to this, but Brave and Opera can. We chose to implement this and I've also kicked off an issue in the permissions repo at W3C to see if browsers are open to extending this API to extensions.

Reply:
FWIW we added this attribute only because we ran into real sites that were using 3p iframes for features that required `window.ethereum` access. IMO solving clickjacking is not in scope here - if a site doesn't want to be framed, it should use CSP or x-frame-options to signal so to prevent clickjacking.

Reply:
this follows the same security model as https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-allow